year18/src/day10.rs

```diff
@@ -1,10 +1,315 @@
+
+    fn north(&self) -> i32 {
+        self.stars
+            .iter()
+            .map(|star| star.p.y)
+            .max()
+            .unwrap()
+    }
+
+    fn east(&self) -> i32 {
+        self.stars
+            .iter()
+            .map(|star| star.p.x)
+            .min()
+            .unwrap()
+    }
+
+    fn south(&self) -> i32 {
+        self.stars
+            .iter()
+            .map(|star| star.p.y)
+            .min()
+            .unwrap()
+    }
+
+    fn west(&self) -> i32 {
+        self.stars
+            .iter()
+            .map(|star| star.p.x)
+            .max()
+            .unwrap()
+    }
+
+    fn height(&self) -> u32 {
+        self.north().abs_diff(self.south()) + 1
+    }
+
+    fn width(&self) -> u32 {
+        self.east().abs_diff(self.west()) + 1
+    }
+
+    fn northeast(&self) -> Vector2<i32> {
+        Vector2::new(self.east(), self.north())
+    }
+
+    // The `frame` is the key object here. It's a u8 (0-255) array of size 4 * width * height.
+    fn draw(&self, frame: &mut [u8], height: u32, width: u32, northeast: Vector2<i32>) {
+        // Each pixels is 4 elements of the array—rbga format.
+        for (i, pixel) in frame.chunks_exact_mut(4).enumerate() {
+            
+            // The frame coordinate system, S', has the origin at the top left corner,
+            // x increases from left to right and y increases from top to bottom.
+            let p_prime = Vector2::new(
+                i as i32 % width as i32,
+                height as i32 - 1 - i as i32 / width as i32
+            );
+
+            // The sky coordinate system, S, has the origin somewhere in the middle.
+            // The northeast-most point in the sky is at the top left.
+            // x increases from left to right and y increases from bottom to top.
+            // To convert the point p' to this coordinate system involves offsetting by
+            // the northeast point and flipping about the X axis.
+            let p: Vector2<i32> = MIRROR_X_AXIS * p_prime + northeast;
+
+            let mut rgba = BACKGROUND;
+            for star in &self.stars {
+                if star.p == p {
+                    rgba = star.colour;
+                    break
+                }
+            }
+
+            pixel.copy_from_slice(&rgba);
+        }
+    }
+}
+
+pub fn day10(input_lines: &[Vec<String>]) -> (String, String) {
+    let stars: Vec<Star> = input_lines[0]
+        .iter()
+        .map(|line| line.parse::<Star>().unwrap())
+        .collect();
+
+    let mut sky = NightSky::new(stars);
+
+    // Minimize the height of the sky.
+    // This isn't guaranteed to work, but it does for my case.
+    // I implemented the interactive mode below so that a user could step forwards and backwards
+    // through the animation to find the right time.
+    // Without first reducing the height of the sky, however, the frame would have been too large.
+    let mut height = sky.height();
+    while sky.height() <= height {
```

Comment:
Ah so this updates the set of stars until the area starts growing again? I'm wondering if there's a more efficient way to write this, sky.height() loops through all of the particles twice and sky.update() does a third time - there may well not be (and I think it's more efficient than mine!), I really want to know if this problem can be solved without applying some test to all of the particles in the set every iteration